database/scrapers/event_calendar.ts

```diff
@@ -0,0 +1,230 @@
+import { DateTime } from "luxon";
+
+import { BaseScraperModule, TaskHandle } from "#commands/db_scrape";
+import CalendarEvent from "#models/calendar_event";
+import { ICSRoot, ICSValue, parseICS } from "#utils/ics";
+
+const FETCH_LIMIT_DAYS = 30;
+const CALENDAR_ID = "9ke30hbjjke60u5jbii42g2rpo@group.calendar.google.com";
+const FIELD_LABELS = {
+  startDate: "DTSTART",
+  endDate: "DTEND",
+  googleCalId: "UID",
+  desc: "DESCRIPTION",
+  location: "LOCATION",
+  name: "SUMMARY",
+};
+
+interface GoogleCalendarEventDto {
+  googleCalId: string;
+  startTime: DateTime<true>;
+  endTime: DateTime<true>;
+  name: string;
+  location: string | null;
+  description: string | null;
+}
+
+class CalendarParser {
+  private events: ICSValue[];
+
+  constructor(
+    private calendarString: string,
+    private task: TaskHandle,
+  ) {
+    if (
+      !this.calendarString.startsWith("BEGIN:VCALENDAR") ||
+      !this.calendarString.trimEnd().endsWith("END:VCALENDAR")
+    ) {
+      throw new Error("Calendar invalid");
+    }
+    const calendarRoot = parseICS(calendarString, (message) =>
+      task.update(message),
+    ).VCALENDAR;
+    if (calendarRoot === undefined) {
+      throw new Error("Calendar invalid - missing calendar root");
+    }
+    const eventArray = (calendarRoot as ICSRoot).VEVENT;
+    if (eventArray === undefined || !Array.isArray(eventArray)) {
+      throw new Error("Calendar invalid - missing event array");
+    }
+    this.events = eventArray;
+  }
+
+  public getLastEvent(): GoogleCalendarEventDto | null {
+    let event = this.extractLastEvent();
+    while (event === undefined) {
+      event = this.extractLastEvent();
+    }
+    return event;
+  }
+
+  //null - none left, undefined - malformed event
+  private extractLastEvent(): GoogleCalendarEventDto | null | undefined {
+    if (this.events.length === 0) {
+      return null;
+    }
+    const eventBlock = this.events.pop();
+    if (eventBlock === undefined) {
+      return undefined;
+    }
```

Comment:
`this.events.length === 0` and `this.events.pop() === undefined` are basically the same check